crates/oci-distribution/src/client.rs

```diff
@@ -114,6 +127,73 @@ impl Client {
         })
     }
 
+    /// Push an image and return the uploaded URL of the image
+    ///
+    /// The client will check if it's already been authenticated and if
+    /// not will attempt to do.
+    ///
+    /// If a manifest is not provided, the client will attempt to generate
+    /// it from the provided image and config data.
+    ///
+    /// Returns pullable URL for the image
+    pub async fn push(
+        &mut self,
+        image_ref: &Reference,
+        image_data: &ImageData,
+        config_data: &[u8],
+        config_media_type: &str,
+        auth: &RegistryAuth,
+        manifest: Option<OciManifest>,
+    ) -> anyhow::Result<String> {
+        debug!("Pushing image: {:?}", image_ref);
+
+        if !self.tokens.contains_key(image_ref.registry()) {
+            self.auth(image_ref, auth, &RegistryOperation::Push).await?;
+        }
+
+        // Start push session
+        let mut location = self.begin_push_session(image_ref).await?;
+
+        // Upload layers
+        let mut start_byte = 0;
+        for layer in &image_data.layers {
+            // Destructuring assignment is not yet supported
+            let (next_location, next_byte) = self
+                .push_layer(&location, &image_ref, layer.data.to_vec(), start_byte)
+                .await?;
```

Comment:
Yes we aren't doing any cleanup here, taking a look at https://github.com/opencontainers/distribution-spec/blob/master/spec.md#content-management it seems to say that registries _can_ support deletion, but might not 🤔  I think ideally it should get cleaned up (during my tested I failed pushes _multiple_ times and was able to overwrite the blobs) but it would be moot if the registry had deletion disabled

Reply:
That being said cleanup on failure doesn't appear to be a large change, so I'd be comfortable implementing it or leaving it as a TODO.